pallets/liquidity-pools-gateway/src/lib.rs

```diff
@@ -305,65 +456,58 @@ pub mod pallet {
 				None => Err(Error::<T>::MessagePackingNotStarted.into()),
 			}
 		}
-	}
-
-	impl<T: Config> Pallet<T> {
-		/// Give the message to the `InboundMessageHandler` to be processed.
-		fn process_inbound_message(
-			domain_address: DomainAddress,
-			message: T::Message,
-		) -> (DispatchResult, Weight) {
-			let mut count = 0;
-
-			for submessage in message.submessages() {
-				count += 1;
-
-				if let Err(e) = T::InboundMessageHandler::handle(domain_address.clone(), submessage)
-				{
-					// We only consume the processed weight if error during the batch
-					return (Err(e), LP_DEFENSIVE_WEIGHT.saturating_mul(count));
-				}
-			}
-
-			(Ok(()), LP_DEFENSIVE_WEIGHT.saturating_mul(count))
-		}
 
-		/// Retrieves the router stored for the provided domain, sends the
-		/// message using the router, and calculates and returns the required
-		/// weight for these operations in the `DispatchResultWithPostInfo`.
-		fn process_outbound_message(
-			sender: DomainAddress,
+		/// Manually increase the proof count for a particular message and
+		/// executes it if the required count is reached.
+		///
+		/// Can only be called by `AdminOrigin`.
+		#[pallet::weight(T::WeightInfo::execute_message_recovery())]
+		#[pallet::call_index(11)]
+		pub fn execute_message_recovery(
+			origin: OriginFor<T>,
 			domain: Domain,
-			message: T::Message,
-		) -> (DispatchResult, Weight) {
-			let router_ids = T::RouterId::for_domain(domain);
-
-			// TODO handle router ids logic
+			proof: Proof,
+			router_id: T::RouterId,
+		) -> DispatchResult {
+			T::AdminOrigin::ensure_origin(origin)?;
 
-			let mut count = 0;
-			let bytes = message.serialize();
+			let session_id = InboundMessageSessions::<T>::get(&domain)
+				.ok_or(Error::<T>::InboundDomainSessionNotFound)?;
 
-			for router_id in router_ids {
-				count += 1;
-				if let Err(e) = T::MessageSender::send(router_id, sender.clone(), bytes.clone()) {
-					return (Err(e), LP_DEFENSIVE_WEIGHT.saturating_mul(count));
-				}
-			}
+			let routers = Routers::<T>::get(&domain).ok_or(Error::<T>::RoutersNotFound)?;
 
-			// TODO: Should we fix weights?
-			(Ok(()), LP_DEFENSIVE_WEIGHT.saturating_mul(count))
```

Comment:
We can think later if we need a more accurate weight for outbound. I think for now it's ok having this. The previous version skipped some weights in the returned value

Reply:
By now, the gateway is in charge of putting defensive weights in both inbound and outbound directions. So we can have centralized the weight management until polishing it

Reply:
@wischli also interested in your thoughts

Reply:
> By now, the gateway is in charge of putting defensive weights in both inbound and outbound directions. So we can have centralized the weight management until polishing it

Sounds perfect! 

> We can think later if we need a more accurate weight for outbound. I think for now it's ok having this. The previous version skipped some weights in the returned value

For now, seems fine. I think the weight is just noise at this point for the audit. Can iterate over that once everything is settled.